pkg/tsdb/postgres/postgres.go

```diff
@@ -17,20 +23,38 @@ import (
 )
 
 func init() {
-	tsdb.RegisterTsdbQueryEndpoint("postgres", newPostgresQueryEndpoint)
+	registry.Register(&registry.Descriptor{
+		Name:         "PostgresService",
+		InitPriority: registry.Low,
+		Instance:     &postgresService{},
+	})
 }
 
-func newPostgresQueryEndpoint(datasource *models.DataSource) (tsdb.TsdbQueryEndpoint, error) {
-	logger := log.New("tsdb.postgres")
-	logger.Debug("Creating Postgres query endpoint")
+type postgresService struct {
+	Cfg *setting.Cfg `inject:""`
 
-	cnnstr, err := generateConnectionString(datasource, logger)
+	mtx    sync.Mutex
+	logger log.Logger
+}
+
+func (s *postgresService) Init() error {
+	s.logger = log.New("tsdb.postgres")
+	tsdb.RegisterTsdbQueryEndpoint("postgres", func(ds *models.DataSource) (tsdb.TsdbQueryEndpoint, error) {
+		return s.newPostgresQueryEndpoint(ds)
+	})
+	return nil
+}
+
+func (s *postgresService) newPostgresQueryEndpoint(datasource *models.DataSource) (tsdb.TsdbQueryEndpoint, error) {
+	s.logger.Debug("Creating Postgres query endpoint")
+
+	cnnstr, err := s.generateConnectionString(datasource)
 	if err != nil {
 		return nil, err
 	}
 
 	if setting.Env == setting.Dev {
```

Comment:
Migrating away from using global settings

```suggestion
	if s.Cfg.Env == setting.Dev {
```